src/reflect/scala/reflect/internal/Definitions.scala

```diff
@@ -813,16 +843,27 @@ trait Definitions extends api.StandardDefinitions {
      * The class defining the method is a supertype of `tp` that
      * has a public no-arg primary constructor.
      */
-    def samOf(tp: Type): Symbol = if (!settings.Xexperimental) NoSymbol else findSam(tp)
-
-    def findSam(tp: Type): Symbol = {
-      // if tp has a constructor, it must be public and must not take any arguments
-      // (not even an implicit argument list -- to keep it simple for now)
-      val tpSym  = tp.typeSymbol
-      val ctor   = tpSym.primaryConstructor
-      val ctorOk = !ctor.exists || (!ctor.isOverloaded && ctor.isPublic && ctor.info.params.isEmpty && ctor.info.paramSectionCount <= 1)
+    def samOf(tp: Type): Symbol = {
+      // look at erased type because we (only) care about what ends up in bytecode
+      // (e.g., an alias type or intersection type is fine as long as the intersection dominator compiles to an interface)
+      val tpSym = erasure.javaErasure(tp).typeSymbol
+
+      if (tpSym.exists
+          // We use Java's MetaLambdaFactory, which requires an interface for the sam's owner
+          // (TODO: Can't use isInterface, yet, as it hasn't been updated for the new trait encoding)
+          && (tpSym.isJavaInterface || tpSym.isTrait)
```

Comment:
The original implementation allowed classes:

```
⚡ scala -Xexperimental
Welcome to Scala 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_65).
Type in expressions for evaluation. Or try :help.

scala> abstract class SamClass { def apply(a: Any): Any }
defined class SamClass

scala> def foo: SamClass = (x: Any) => x
foo: SamClass

scala> :quit
```

We exclude them now:

```
scala> abstract class SamClass { def apply(a: Any): Any }
defined class SamClass

scala> def foo: SamClass = (x: Any) => x
<console>:12: error: type mismatch;
 found   : Any => Any
 required: SamClass
       def foo: SamClass = (x: Any) => x
                                    ^
```

Same for non-static traits:

```
⚡ scala -Xexperimental
Welcome to Scala 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_65).
Type in expressions for evaluation. Or try :help.

scala> :paste -raw
// Entering paste mode (ctrl-D to finish)

package p { class C { trait T { def a(a: Any): Any } } }

// Exiting paste mode, now interpreting.

scala> val c = new p.C
c: p.C = p.C@de96c95

scala> def foo: c.T = (a: Any) => a
foo: c.T
```

So the restrictions here aren't inherent, they are a design choice to only allow SAMs that can represented through LMF.

I actually think there is some merit to this restriction, there is likely to be less WTF-s for "accidental SAMs" like `InputStream`

```
scala> (() => 42): java.io.InputStream
res0: java.io.InputStream = $anonfun$1@756bc363

```

But it is worth noting prominently and discussing.